File: numerical_integration.py
import LaplacesDemon
import matrixStats
import cubature
import Rcpp
from scipy.stats import truncexpon, poisson, gamma
import numpy as np

# Reading relevant functions and simulation settings
# source(paste0(getwd(), "/kernel_functions.R"))
# source(paste0(getwd(), "/simulation_functions.R"))

# Reading the emcee posterior sample
emcee_results_list = LaplacesDemon.readRDS(file="/simple/emcee/results/emcee_result.RDS")
print(type(emcee_results_list$posterior_sample))

################
# Posterior kernel
################

def t_Xlik(x, t_xi, A, Time):
    xi = np.exp(t_xi)
    return poisson.logpmf(x, A * Time * xi)

def t_Lambda_ker(y, Lambda, t_mu, t_theta, t_pid, t_xi, A, Time, a, r, e):
    mu = np.exp(t_mu)
    theta = np.exp(t_theta)
    pid = LaplacesDemon.invlogit(t_pid)
    xi = np.exp(t_xi)

    log_likelihood = poisson.logpmf(y, (a * xi) + (r * e * Lambda) * Time)

    jacobian = np.log(pid) + LaplacesDemon.logdeltafunction(Lambda) + \
                np.log(1 - pid) + gamma.logpdf(Lambda, a=np.exp(2 * t_mu - t_theta), scale=np.exp(t_mu - t_theta))

    return log_likelihood + jacobian

def t_Lambda_gamma(y, Lambda, t_mu, t_theta, t_xi, A, Time, a, r, e):
    lshape = 2 * t_mu - t_theta
    lrate = t_mu - t_theta
    xi = np.exp(t_xi)

    lprior = gamma.logpdf(Lambda, a=np.exp(lshape), scale=np.exp(lrate))
    llike = poisson.logpmf(y, (a * xi) + (r * e * Lambda) * Time)

    return lprior + llike

#################
# Numerical integration
#################

def numerical(t_mu, t_theta, t_pid, t_xi, Lambda, x, y, A, Time, a, r, e):
    pid = LaplacesDemon.invlogit(t_pid)
    xi = np.exp(t_xi)

    o = np.array([t_Lambda_gamma(y, Lambda[Lambda != 0], t_mu, t_theta, t_xi, A, Time, a, r, e) for Lambda_i in Lambda[Lambda != 0]])
    io = matrixStats.logsumexp(o, axis=1) - np.log(o.shape[1])

    zeroinf = np.log(pid) + poisson.logpmf(y, lambda=a * xi * Time)
    nonzeroinf = np.log(1 - pid) + io

    return t_Xlik(x, t_xi, A, Time) + np.sum(np.vectorize(matrixStats.logsumexp)(zeroinf, nonzeroinf))

# Example usage
numerical(np.log(mu_star), np.log(theta_star), t_pid, np.log(xi_star),
          np.arange(0, 7, 0.1),
          x=30, y=np.arange(0, 3), A=A, Time=Time, a=a, r=R, e=e)
numerical(t_mu, t_theta, t_pid, t_xi,
          np.arange(0, 1, 1e-5),
          x=30, y=np.arange(0, 25), A=A, Time=Time, a=a, r=R, e=e)

##########################
# Analytical solution
##########################

def analytical(t_mu=None, t_theta=None, t_shape=2 * t_mu - t_theta,
               t_rate=t_mu - t_theta, t_pid=None, t_xi=None,
               D=None, n=None, A=None, Time=None, a=None, r=None, e=None):
    X = D[n + 1]
    Y = D[0:n]

    def summand(yi, k, t_shape, t_rate, t_xi, Time, a, r, e):
        lcombination = np.sum([LaplacesDemon.lfactorial(yi) - LaplacesDemon.lfactorial(k) - LaplacesDemon.lfactorial(yi - k)])
        lpowers = (yi - k) * (np.log(a) + t_xi + np.log(Time)) + k * (np.log(r) + np.log(e) + np.log(Time))
        llaplace = gamma.logpdf(np.exp(t_shape) + k, a=np.exp(t_shape), scale=-np.exp(t_shape - t_rate - np.log(r) - np.log(e) - np.log(Time)))
        return lcombination + lpowers + llaplace

    def prodant(yi, t_shape, t_rate, t_xi, t_pid, Time, a, r, e):
        pid = LaplacesDemon.invlogit(t_pid)
        lpois = -a * np.exp(t_xi) * Time - LaplacesDemon.lfactorial(yi)
        l0inf = yi * (np.log(a) + t_xi + np.log(Time)) + np.log(pid)
        sums = np.log(1 - pid) + np.exp(t_shape) * (t_rate) - gamma.logpdf(np.exp(t_shape)) + \
               LaplacesDemon.logsumexp(np.vectorize(summand)(yi, np.arange(0, yi + 1), np.full_like(t_shape, t_shape),
                                                            np.full_like(t_rate, t_rate), np.full_like(t_xi, t_xi),
                                                            np.full_like(Time, Time), np.full_like(a, a), np.full_like(r, r),
                                                            np.full_like(e, e)), axis=1)
        return lpois + LaplacesDemon.logplus(l0inf, sums)

    def prodant_vec(Y, t_shape, t_rate, t_xi, t_pid, Time, a, r, e):
        return np.vectorize(prodant)(Y, np.full_like(t_shape, t_shape), np.full_like(t_rate, t_rate),
                                     np.full_like(t_xi, t_xi), np.full_like(t_pid, t_pid), np.full_like(Time, Time),
                                     np.full_like(a, a), np.full_like(r, r), np.full_like(e, e))

    return poisson.logpmf(X, A * Time * np.exp(t_xi)) + np.sum(prodant_vec(Y, t_shape, t_rate, t_xi, t_pid, Time, a, r, e))

# Example usage
analytical(t_shape=t_shape, t_rate=t_rate, t_xi=t_xi, t_pid=t_pid,
           D=emcee_results_list$simulated_data, n=len(emcee_results_list$simulated_data) - 1,
           A=A, Time=Time, a=a, r=R, e=e)

################
# Example of numerical methods accuracy
################

# Debug testing: integrate()
def test_integral(Lambda, t_mu, t_theta, r, e, Time):
    mu = np.exp(t_mu)
    theta = np.exp(t_theta)
    shape = mu ** 2 / theta
    return np.exp((shape - 1) * np.log(Lambda) - Lambda * shape - r * e * Time * Lambda)

i = cubature.integrate(test_integral, np.array([0]), np.array([np.inf]), args=(t_mu, t_theta, R, e, Time))
print(np.log(i[0]))

# Numerical integration
def l_test_integral(t_mu, t_theta, t_pid, t_xi, Lambda, x, y, A, Time, a, r, e):
    pid = LaplacesDemon.invlogit(t_pid)
    xi = np.exp(t_xi)

    o = np.array([t_Lambda_gamma(y, Lambda[Lambda != 0], t_mu, t_theta, t_xi, A, Time, a, r, e) for Lambda_i in Lambda[Lambda != 0]])
    io = matrixStats.logsumexp(o, axis=1) - np.log(o.shape[1])

    zeroinf = np.log(pid) + poisson.logpmf(y, lambda=a * xi * Time)
    nonzeroinf = np.log(1 - pid) + io

    return np.vectorize(LaplacesDemon.logplus)(zeroinf, nonzeroinf)

l_test_integral(t_mu, t_theta, t_pid, t_xi, np.arange(0, 1e-0, 1e-7),
                x=emcee_results_list$simulated_data[len(emcee_results_list$simulated_data) - 1],
                y=emcee_results_list$simulated_data[0:len(emcee_results_list$simulated_data) - 1],
                A=A, Time=Time, a=a, r=R, e=e)

# Plotting
# plot(seq(0, 5e-4, by = 1e-8), exp(l_test_integral(seq(0, 5e-4, by = 1e-8), t_mu, t_theta, R, e, Time)))
# The peak is around 7e-5
# True value via Laplace:
# test_inteana <- function(k = 0, t_mu, t_theta, r, e, Time) {
#   mu <- exp(t_mu)
#   theta <- exp(t_theta)
#   shape <- mu^2 / theta
#
#   ifelse((shape + k) > 0, lgamma(shape) + (-shape) * log(shape + r * e * Time), -Inf)
# }
# test_inteana(t_mu = t_mu, t_theta = t_theta, r = R, e = e, Time = Time)
# LESSON: Focus on the peak and don't include delta functions (that changes the weighting of the whole integral).

##################
# Complete integrated kernel (partial marginal likelihood)
##################
# Remaining code for the integrated kernel is not provided.

File: R_samplers/numerical_integration.R
library(LaplacesDemon) # for logistic function
library(matrixStats)
library(cubature)
library(Rcpp)
sourceCpp("logsum.cpp")

# reading relevant functions and simulation settings
#source(paste0(getwd(), "/kernel_functions.R"))
#source(paste0(getwd(), "/simulation_functions.R"))

# reading the emcee posterior sample
emcee_results_list <- readRDS(file = paste0(getwd(), "/simple/emcee/results/emcee_result.RDS"))
class(emcee_results_list$posterior_sample)

################
# posterior kernel
###############
t_Xlik <- function(x, t_xi, A, Time) {
  xi <- exp(t_xi)
  dpois(x = x,
        A * Time * xi, 
        log = TRUE)
}

t_Lambda_ker <- function(y, Lambda, t_mu, t_theta, t_pid, t_xi,
                         A, Time, a, r, e) {
  #Lambda <- exp(t_Lambda)
  mu <- exp(t_mu)
  theta <- exp(t_theta)
  pid <- invlogit(t_pid)
  xi <- exp(t_xi)
  
  dpois(x = y, 
        (c(a * xi) + c(r) * c(e) * Lambda) * c(Time), 
        log = TRUE) + #Jacobian:
    #t_Lambda + # because Lambda is fed into the function, not log Lambda, so the Jacobian is not needed.
    sapply(Lambda, 
           function(i) logplus(log(pid) + logdeltafunction(i),
                               log(1 - pid) + dgamma(x = i,
                                                     shape = exp(2 * t_mu - t_theta),
                                                     rate = exp(t_mu - t_theta),
                                                     log = TRUE)))
}
# plotting
#plot(seq(0, 5e-3, by = 1e-7), exp(t_Lambda_ker(200, seq(0, 5e-3, by = 1e-7), t_mu, t_theta, t_pid, t_xi, A, Time, a, R, e))) # peak depends on data.

t_Lambda_gamma <- function(y, Lambda, t_mu, t_theta, t_xi, A, Time, a, r, e) {
  lshape <- 2 * t_mu - t_theta
  lrate <- t_mu - t_theta
  xi <- exp(t_xi)
  
  lprior <- dgamma(Lambda, shape = exp(lshape), rate = exp(lrate), log = TRUE)
  llike <- dpois(x = y, 
                 (c(a * xi) + c(r) * c(e) * Lambda) * c(Time), 
                 log = TRUE)
  lprior + llike
}

#########
# numerical integration
#########
numerical <- function(t_mu, t_theta, t_pid, t_xi, Lambda, # Lambda > 0 needed
                      x, y, A, Time, a, r, e) {
  pid <- invlogit(t_pid)
  xi <- exp(t_xi)
  
  o <- outer(y, Lambda[Lambda != 0], FUN = t_Lambda_gamma,
             t_mu = t_mu, t_theta = t_theta, t_xi = t_xi, A = A, Time = Time, a = a, r = r, e = e)
  io <- apply(o, 1, logplusvec) - log(ncol(o))
  zeroinf <- log(pid) + dpois(y, lambda = a * xi * Time, log = TRUE)
  nonzeroinf <- log(1 - pid) + io
  t_Xlik(x, t_xi, A, Time) + sum(sapply(1:length(zeroinf), function(i) logplus(zeroinf[i], nonzeroinf[i])))
  
  
  # test
  #list(t_Xlik(x, t_xi, A, Time),
  #     io)
}


##########################
# analytical solution
#########################
analytical <- function(t_mu = NULL, t_theta = NULL, 
                       t_shape = 2 * t_mu - t_theta, 
                       t_rate = t_mu - t_theta, t_pid, t_xi, 
                       D, n = length(D) - 1, A, Time, a, r, e) {
  # reshape formal arguments
  X <- D[n + 1]
  Y <- D[1:n]
  
  # construction
  summand <- function(yi, k, t_shape, t_rate, t_xi, Time,
                      a, r, e) {
    lcombination <- lfactorial(yi) - (lfactorial(k) + lfactorial(yi-k))
    lpowers <- (yi - k) * (log(a) + t_xi + log(Time)) + k * (log(r) + log(e) + log(Time))
    llaplace <- lgamma(exp(t_shape) + k) + (-exp(t_shape) - k) * (logplus(t_rate, log(r) + log(e) + log(Time)))
    lcombination + lpowers + ifelse((exp(t_shape) + k) > 0, llaplace, -Inf) # the condition of Laplace transform
  }
  prodant <- function(yi, t_shape, t_rate, t_xi, t_pid, Time,
                      a, r, e) {
    pid <- invlogit(t_pid)
    
    lpois <- -a * exp(t_xi) * Time - lfactorial(yi)
    l0inf <- yi * (log(a) + t_xi + log(Time)) + log(pid)
    sums <- log(1 - pid) + exp(t_shape) * (t_rate) - lgamma(exp(t_shape)) + 
      logplusvec(summand(yi = yi, k = 0:yi, t_shape = t_shape, t_rate = t_rate, t_xi = t_xi, Time = Time, a = a, r = r, e = e))
    lpois + logplus(l0inf, sums)
    
    ############## test
    #dpois(yi, lambda = a * exp(t_xi) * Time, log = TRUE) + log(pid)
  }
  prodant_vec <- function(Y, t_shape, t_rate, t_xi, t_pid, Time, 
                          a, r, e) {
    sapply(Y, prodant, t_shape = c(t_shape), t_rate = c(t_rate), t_pid = c(t_pid), t_xi = c(t_xi), Time = c(Time), a = c(a), r = c(r), e = c(e))
  }
  dpois(x = X, lambda = A * Time * exp(t_xi), log = TRUE) + 
    sum(prodant_vec(Y = Y, t_shape = t_shape, t_rate = t_rate, t_pid = t_pid, t_xi = t_xi, Time = Time, a = a, r = r, e = e))
  # test
  #list(dpois(x = X, lambda = A * Time * exp(t_xi), log = TRUE),
  #     prodant_vec(Y = Y, t_mu = t_mu, t_theta = t_theta, t_pid = t_pid, t_xi = t_xi, Time = Time, a = a, r = r, e = e))
}


##################
# complete integrated kernel (partial marginal likelihood)
#################